File: backend/app/services/road_polyline.py
"""Bounded geometry codec, independent of databases and application secrets."""


def decode_road_geometry(encoded):
    """Decode polyline6 into GeoJSON longitude/latitude pairs."""
    if not isinstance(encoded, str) or not 1 <= len(encoded) <= 2_000_000:
        raise ValueError('invalid_road_document_geometry')
    index = latitude = longitude = 0
    points = []

    def delta():
        nonlocal index
        value = shift = 0
        while True:
            if index >= len(encoded) or shift > 30:
                raise ValueError('invalid_road_document_geometry')
            byte = ord(encoded[index]) - 63
            index += 1
            if not 0 <= byte <= 63:
                raise ValueError('invalid_road_document_geometry')
            value += (byte & 31) * 2 ** shift
            if byte < 32:
                return -(value // 2 + 1) if value % 2 else value // 2
            shift += 5

    while index < len(encoded):
        latitude += delta()
        longitude += delta()
        if abs(latitude) > 85_000_000 or abs(longitude) > 180_000_000 or len(points) >= 100000:
            raise ValueError('invalid_road_document_geometry')
        points.append([longitude / 1e6, latitude / 1e6])
    if len(points) < 2:
        raise ValueError('invalid_road_document_geometry')
    return points

File: frontend/src/components/Map/roadPolyline.ts
/** Decode the fixed engine's polyline6 into Leaflet [latitude, longitude]. */
export function roadPolyline(encoded: string): Array<[number, number]> {
  if (!encoded || encoded.length > 2_000_000) throw new Error('无效参考路径')
  let index = 0, latitude = 0, longitude = 0
  const points: Array<[number, number]> = []
  const delta = () => {
    let value = 0, shift = 0
    while (true) {
      if (index >= encoded.length || shift > 30) throw new Error('路径编码不完整')
      const byte = encoded.charCodeAt(index++) - 63
      if (byte < 0 || byte > 63) throw new Error('无效路径字符')
      value += (byte & 31) * 2 ** shift
      if (byte < 32) break
      shift += 5
    }
    return value % 2 ? -(Math.floor(value / 2) + 1) : value / 2
  }
  while (index < encoded.length) {
    latitude += delta(); longitude += delta()
    if (Math.abs(latitude) > 85_000_000 || Math.abs(longitude) > 180_000_000 || points.length >= 100000) throw new Error('路径超出有效范围')
    points.push([latitude / 1e6, longitude / 1e6])
  }
  if (points.length < 2) throw new Error('路径点不足')
  return points
}
